### DISCUSSION TITLE
原点復帰コマンドについて

### DISCUSSION THREAD

**Discussion by Takamatsu-Matrix** [ponoor/step-series-support#16] [Q&A] [answered]
装置に原点センサーをつけ、原点復帰動作をさせようとしています。

device.set(commands.SetHomingDirection(Axis_X,True))
device.set(commands.SetHomingSpeed(Axis_X,100))
device.set(commands.Homing(Axis_X))

と記述したのですが、モーターがコトコト音を立ててゆっくり刻むように動きます。
あまりに遅いのでタイムアウトしてしまいます。

そこでこのSetHomingSpeed(Axis_X,100)の部分を100をそれ以上にあげても、速度は変わりません。
仮に10000にセットしても、同じ動作をします。

なにか間違いがあるのでしょうか？？
原点復帰の速度を変えたいのですが、教えていただければ助かります。

**Comment by kanta** [comment]
@Takamatsu-Matrix いくつか原因が考えられそうです。
まず、原点復帰の速さが本当に変更できているか、以下のコードで確認してみてください。
```Python
from stepseries import commands, responses, step400

# Default handler for all responses
def default_handler(message: responses.OSCResponse) -> None:
    print("Message received:", message)

if __name__ == '__main__':
    # Configurations that should be changed
    dip_switch_id = 1  # Should match what is set on the device
    local_ip_address = "10.0.0.101"  # The ip address of the device
    local_port = 50000
    server_address = "0.0.0.0"  # The address of the server; should always be 0.0.0.0 (the local machine)
    server_port = 50101

    device = step400.STEP400(dip_switch_id, local_ip_address, local_port, server_address, server_port)

    # Register a default handler for all responses
    device.on(None, default_handler)

    device.set(commands.SetDestIP())

    # Change the homing direction, depends on the physical setting.
    device.set(commands.SetHomingDirection(1, False))
    device.set(commands.SetHomingSpeed(1,400.0))
    device.get(commands.GetHomingSpeed(1))
    device.set(commands.SetHomingSpeed(1,200.0))
    device.get(commands.GetHomingSpeed(1))

    #device.set(commands.Homing(1))
```

問題なく設定を変更できていれば、以下のような返答があるのではと思います。
```
Message received: HomingSpeed(address='/homingSpeed', motorID=1, homingSpeed=400.0)
Message received: HomingSpeed(address='/homingSpeed', motorID=1, homingSpeed=200.0)
```

次に、コトコト音を立てているということで、もしかするとKVALの設定が低すぎて脱調気味になっている可能性があります。初期値はいずれも16になっていますので、例えば少し上げて
```Python
device.set(commands.SetKval(1, 10, 24, 24, 24))
```
としてみるなど、モータの定格上で問題のない範囲でKVALを調整してみるのはどうでしょうか。

最後に、原点復帰を含めたすべてのモータ速さは、スピードプロファイルの maxSpeed でリミットされます。もしかするとこの数値が低い値に設定されていないでしょうか。Pythonライブラリでは `commands.GetSpeedProfile` で現在の設定を確認できます。また `commands.SetSpeedProfile` で新しい値をセットできます。

**Reply by Takamatsu-Matrix** [reply]
@kanta 

試してみました。

同じ応答があります。
正しくセットされているようです。

また、このコマンドの前で
device.set(commands.SetKval(Axis_X,0,255,64,64))
を実施してるので、あがってると思います。

また、SetKvalの直後に、
device.set(commands.SetSpeedProfile(Axis_X,2000,2000,5000))
を入れています。

でもダメなようです。

**Comment by kanta** [comment, marked as answer]
@Takamatsu-Matrix ほかに考えられる理由としては、、
[こちらの説明](https://ponoor.com/docs/step-series/functional-description/homing-and-position-management/#i-4)の通り、原点復帰は `GoUntil` と `ReleaseSw` の2ステップからなっていますが、何らかの理由でGoUntil部分がスキップされてReleaseSw部分が実行されている可能性があります。
以下のコードで、homingStatusがコンソールに表示されると思いますので、ゆっくり動いている間は homingStatusが何番になっているか確認してみていただけますか？
このコードでは、homingを開始したあと、ステータスを表示させるためにBUSY状態が終わるまで待つようになっています。

```Python
import time
from urllib import response
from stepseries import commands, responses, step400
busyState = 1

# Default handler for all responses
def default_handler(message: responses.OSCResponse) -> None:
    print("Message received:", message)

# Handler for GetBusy
def busy_handler(message: responses.Busy) -> None:
    global busyState
    busyState = message.state
    motorId = message.motorID

if __name__ == '__main__':
    # Configurations that should be changed
    dip_switch_id = 1  # Should match what is set on the device
    local_ip_address = "10.0.0.101"  # The ip address of the device
    local_port = 50000
    server_address = "0.0.0.0"  # The address of the server; should always be 0.0.0.0 (the local machine)
    server_port = 50101

    device = step400.STEP400(dip_switch_id, local_ip_address, local_port, server_address, server_port)

    # Register a default handler for all responses
    device.on(None, default_handler)

    device.set(commands.SetDestIP())
    device.set(commands.EnableBusyReport(1 , True, busy_handler))
    device.set(commands.SetKval(1, 0, 32, 32, 32))
    
    # Change the homing direction, depends on the physical setting.
    device.set(commands.SetHomingDirection(1, False))

    device.set(commands.Homing(1))

    print("Waiting for BUSY state clear")
    while busyState:
        time.sleep(0.01)
    print("BUSY cleared.")
```

正しく動作しているのであれば、まずセンサに向かって動いている間は
```
Message received: HomingStatus(address='/homingStatus', motorID=1, homingStatus=1)
Message received: Busy(address='/busy', motorID=1, state=1)
```
というように表示されているのではと思います。`homingStatus=2` となっている場合は、`GoUnitl`がスキップされて`ReleaseSw`部分が実行されていることを表しています。

**Reply by Takamatsu-Matrix** [reply]
ということは、もしかしてですけど、原点センサーの論理が逆だったりします？
OMRONのEE-SX472を使っていますが、原点センサを遮光すると0Vになるタイプなので良いと思ってましたが、もしかして逆ですか？

**Reply by Takamatsu-Matrix** [reply]
遮光してテストしたら、論理が逆でした。
正しく動きました。

EE-SX672を使用することにします。

ありがとうございました。

**Reply by kanta** [reply]
動いて良かったです！確かに原点復帰を開始する時点で接点が閉じている場合は `GoUntil`を飛ばして最初から `ReleaseSw`の動作に入ります。
STEPシリーズのモータドライバチップは接点が閉じている＝原点センサが反応している、として扱っています。EE-SX472の場合は入光時に接点が閉じるため、センサが通常時遮光されていて、原点ではスリットがあるなどして入光する、という機構でしたら正常動作するかと思います。
